[Feat] 신고화면 UI 구현

ikseong00 opened the pull request:
## Related issue 🛠
- closed #79 

## Work Description 📝
- 기존 xml 코드 제거 -> Compose 코드로 마이그레이션
- New 신고 화면 UI 구현
- 정보 입력 화면 UI 구현

## Screenshot 📸

https://github.com/user-attachments/assets/d2c8c172-9e9c-4ff4-9bb2-0e3a8905d5b7



## Uncompleted Tasks 😅
- [ ] 최초 권한 정보 Datastore 저장/불러오기
- [ ] 품종 리스트 불러와 저장하기
- [ ] 축종 선택 시 품종 결과 초기화
- [ ] 신고 Api 구현
- [ ] 로딩 상태 시 Indiactor 바 삽입

## To Reviewers 📢
이렇게 구현사항이 많을 줄 몰랐습니다... 이슈를 더 잘게 나눴어야 했나봐요
요구사항을 이렇게 만든 걸 반성하네요

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - 실종/목격 신고 UI를 Compose로 전면 재구성: 이미지 캐러셀·AI 판별, 다양한 입력 컴포넌트(버튼, 라운드 입력, 휠픽커, 스페이서, 탑바), 날짜 바텀시트, 이미지/설정 다이얼로그, 지도 기반 위치 선택 등 추가

- Refactor
  - 내비게이션 재구성: 기존 Report 흐름 제거, 실종/목격 신고를 최상위 목적지로 분리; 관련 프래그먼트 레이아웃을 ComposeView로 전환

- Changes
  - 카메라 기능을 선택적 사용으로 전환 및 저장 권한 제거; 털색 옵션 정리(일부 항목 제거, 색상 코드 보강), 성별 옵션 축소 및 관련 데이터 모델 변경

- Chores
  - Naver Map Compose·Accompanist Permissions 의존성 추가 및 빌드 설정 개선
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ikseong00 committed:
[feat] #79 실종 신고 Compose용 Fragment 생성

ikseong00 committed:
[chore] #79 점박이 털 색상 제거

ikseong00 committed:
[feat] #79 Immutable Collect 의존성 추가

ikseong00 committed:
[feat] #79 실종신고 ViewModel, UiState, Event, Effect 구현

ikseong00 committed:
[feat] #79 실종신고 Fragment 기본 세팅 완료

ikseong00 committed:
[feat] #79 실종신고 네비게이션 구현

ikseong00 committed:
[feat] #79 위치 설정 웹뷰 이벤트 구현

ikseong00 committed:
[feat] #79 정보 입력 화면 이동 이벤트 구현

ikseong00 committed:
[fix]: #79 immutable collection 제거

ikseong00 committed:
[refactor]: #79 navigation, screen 패키지로 리패키징

ikseong00 committed:
[feat]: #79 FindUTopAppBar 구현

ikseong00 committed:
[feat]: #79 커스텀 Spacer 구현

ikseong00 committed:
[feat]: #79 실종신고 탑바, 사진 등록 컴포넌트 구현

ikseong00 committed:
[feat]: #79 실종신고 품종 섹션 구현

ikseong00 committed:
[feat]: #79 신고 성별 선택 컴포넌트 구현

ikseong00 committed:
[feat]: #79 사각형 Common Input Field 컴포넌트 구현

ikseong00 committed:
[feat]: #79 RFID 컴포넌트 구현

ikseong00 committed:
[feat]: #79 털색 컴포넌트 구현

ikseong00 committed:
[feat]: #79 날짜 컴포넌트 구현

ikseong00 committed:
[feat]: #79 추가 설명 작성하기 컴포넌트 구현

ikseong00 committed:
[feat]: #79 위치 설정 컴포넌트 구현

ikseong00 committed:
[feat]: #79 실종신고 UI 구현

ikseong00 committed:
[feat]: #79 목격신고 ViewModel 구현

ikseong00 committed:
[feat]: #79 목격신고 UI 구현

ikseong00 committed:
[feat]: #79 목격신고 Fragment 및 네비에기션 설정

ikseong00 committed:
[feat]: #79 신고 이미지 선택 Dialog 구현

ikseong00 committed:
[feat]: #79 카메라, 갤러리 기능 구현

ikseong00 committed:
[feat]: #79 앱 세팅 이동 기능 구현

ikseong00 committed:
[feat]: #79 실종 신고 정보 입력 화면 구현

ikseong00 committed:
[feat]: #79 실종 신고 정보 입력 화면 구현

ikseong00 committed:
[feat]: #79 목격 신고 정보 입력 화면 구현

ikseong00 committed:
[feat]: #79 신고 종료 이벤트 구현

ikseong00 committed:
[feat]: #79 기존 XML 코드 -> Compose 로 마이그레이션

ikseong00 committed:
[fix]: #79 실종/목격동물 게시글 작성 시 이미지 관련 버그 수정

ikseong00 committed:
[chore]: #79 신고 디렉토리 정리

ikseong00 committed:
[feat]: #79 날짜 선택 바텀시트 구현

ikseong00 committed:
[refactor] #79 주소 검색 결과 콜백 등록 로직 onCreate로 이동

ikseong00 committed:
[feat] #87 목격신고 UI 수정 및 기능 추가

- 목격신고 페이지 UI 수정
    - 날짜 선택 BottomSheet 추가
    - 날짜 선택 시 키보드 숨김 처리
    - AI 판별 버튼 추가
    - 기타 UI 수정
- 목격신고 ViewModel 수정
    - 날짜 선택 로직 수정
    - 품종 리스트 초기화 로직 추가
    - AI 판별 이벤트 추가
- 카메라 권한 요청 로직 수정
- 기타 버그 수정

ikseong00 committed:
[choer] #87 Copilot 코드리뷰 반영

ikseong00 committed:
[fix] #87 날짜 선택 바텀시트 동적 날짜 계산 로직 개선

- 하드코딩된 yearList를 동적 범위(2020-2030)로 변경
- monthList를 1-12 전체 월로 수정
- YearMonth.lengthOfMonth() 사용해 정확한 일수 계산
- LaunchedEffect 키에 selectedYear, selectedMonth 추가해 재계산 보장

ikseong00 committed:
[fix] #87 UI 컴포넌트 개선

- InputField placeholder와 텍스트 동시 표시 수정
- WheelPicker startIndex 범위 검증 추가
- AppSettingDialog 하드코딩 텍스트를 string 리소스로 이동
- ReportDescriptionComponent multiline 입력 허용
- 권한 관련 string 리소스 추가

ikseong00 committed:
[feat] #87 신고 화면 기능 개선

- Fragment에 Hilt 의존성 주입 추가
- Fragment lifecycle에서 binding null 처리 추가
- WitnessReportFragment 완료 다이얼로그 타입 수정
- 카메라 네비게이션 로직 개선
- 화면별 위치 업데이트 LaunchedEffect 추가
- WitnessReport 버튼 활성화 조건에 nearPlace 추가

ikseong00 committed:
[fix] #87 ViewModel 및 동물정보 컴포넌트 개선

- 종 선택 시 품종과 검색 텍스트 초기화 로직 추가
- openCamera 메서드 접근제한자 수정
- 다이얼로그 닫기 시 AppSettingDialog 상태도 초기화
- ReportAgeComponent 입력 검증 로직 개선
- ReportBreedComponent placeholder 표시 수정
- ReportSpeciesComponent 동적 segment 계산 수정

ikseong00 committed:
[chore] #87 네비게이션 및 의존성 업데이트

- nav_graph.xml popUpTo 설정 수정으로 네비게이션 스택 관리 개선
- 카메라 화면 네비게이션에서 불필요한 popUpToInclusive 제거
- Map SDK 버전을 3.20.0에서 3.21.0으로 업데이트

ikseong00 committed:
[chore] #87 Code Rabbit 코드리뷰 반영

t1nm1ksun commented on app/src/main/java/com/example/findu/presentation/ui/base/FindUButton.kt:
```diff
@@ -0,0 +1,54 @@
+import com.example.findu.ui.theme.FindUTheme
+
+@Composable
+fun FindUButton(
+    modifier: Modifier = Modifier,
+    @StringRes textRes: Int,
+    onClick: () -> Unit,
+    enabled: Boolean = true,
+) {
```
필수는 아니지만 구글에서 인자 순서를 
1. 필수인자
2. modifier
3. optional인자 
순서를 권장한다고 알고있습니다!
```suggestion
fun FindUButton(
    @StringRes textRes: Int,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
) {
```

ikseong00 replied:
아하 그런 규칙이 있는 줄 몰랐습니다.
https://android.googlesource.com/platform/frameworks/support/+/androidx-main/compose/docs/compose-api-guidelines.md
요 문서에 나와있나요 ??
다시 한 번 공부해야겠습니다..

t1nm1ksun commented on app/src/main/java/com/example/findu/presentation/ui/base/Spacer.kt:
```diff
@@ -0,0 +1,38 @@
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.unit.Dp
+
+@Composable
+fun VerticalSpacer(
+    height: Dp,
+) {
+    Spacer(modifier = Modifier.height(height))
+}
+
+@Composable
+fun HorizontalSpacer(
+    width: Dp,
+) {
+    Spacer(modifier = Modifier.width(width))
+}
```
👍

t1nm1ksun commented on app/src/main/java/com/example/findu/presentation/ui/base/TopAppBar.kt:
```diff
@@ -0,0 +1,80 @@
+    ) {
+        navigationIconRes?.let {
+            IconButton(
+                onClick = { onNavigationIconClick() },
```
```suggestion
                onClick =  onNavigationIconClick,
```
onClick함수를 람다로 감싼다면 호출 시 onNavigationIconClick()을 다시 실행하는 람다 객체가 매번 만들어지기 때문에 불필요한 중첩이 생길수도 있을 것 같습니다!

ikseong00 replied:
[bcc8ac6](https://github.com/FindYou-Kuit/FindYou-Android/pull/82/commits/bcc8ac650d9b956042a54cd728d10db33fac3049) 반영했습니다! , 불필요한 람다 블록이 생기겠네요

t1nm1ksun commented on app/src/main/java/com/example/findu/presentation/ui/common/AppSettingDialog.kt:
```diff
@@ -0,0 +1,82 @@
+                },
+                modifier = Modifier
+                    .fillMaxWidth()
+                    .height(40.dp)
```
컴포넌트의 크기를 고정값으로 정해버린다면(특히 내부에 텍스트가 있을 경우) 사용자가 시스템 설정에서 텍스트 크기를 키웠을 때,
고정값으로 지정해준 높이보다 글씨가 커질 경우 UI가 부자연스러워질 위험이 있다고 생각합니다. 
그래서 저는 보통 padding을 활용해 컴포넌트의 크기를 잡는 편입니다!

<img width="390" height="118" alt="image" src="https://github.com/user-attachments/assets/41710bd4-7e6d-4824-a8f2-f87499622b78" /> 

예를 들어 사진의 버튼을 만들 때 높이를 50으로 주기보다는 

<img width="462" height="112" alt="image" src="https://github.com/user-attachments/assets/1222bc9c-155e-4198-909a-933a7c6fd0ed" />

Text기준으로 padding(vertical=15.dp)를 주는 방식으로요!

제가 제대로 이해하지 못한 걸수도 있으니 혹시 틀린내용이 있다면 말씀해주세요!

ikseong00 replied:
[9dca98b](https://github.com/FindYou-Kuit/FindYou-Android/pull/82/commits/9dca98bc2cabe28d5d790907d07543c98026f344) 반영했습니다!

t1nm1ksun commented on app/src/main/java/com/example/findu/presentation/ui/common/WheelPicker.kt:
```diff
@@ -0,0 +1,138 @@
+    val halfPadItems = (visibleCount - 1) / 2
+
+    // 현재 "가운데"에 가장 가까운 인덱스 계산
+    val currentIndex by remember(
+        state,
+        items
+    ) {
+        derivedStateOf {
+            val offsetItems = state.firstVisibleItemScrollOffset / itemHeightPx
+            val raw = state.firstVisibleItemIndex + offsetItems
+            raw.roundToInt().coerceIn(0, items.lastIndex.coerceAtLeast(0))
+        }
+    }
```
👍

t1nm1ksun commented on app/src/main/java/com/example/findu/presentation/ui/report/component/ReportDateBottomSheet.kt:
```diff
@@ -0,0 +1,247 @@
+                    .padding(vertical = 22.dp)
+            ) {
+                ReportDateBottomSheetContent(
+                    modifier = modifier,
```
구글 가이드라인에 따르면 인자로 받은 modifier는 최상위(root-most) 레이아웃에 적용해야 한다고 합니다!

ikseong00 replied:
[0c855eb](https://github.com/FindYou-Kuit/FindYou-Android/pull/82/commits/0c855eb012725e5c69703b14c3f3b824d2b3185c) 모디파이어 반영했습니다!

t1nm1ksun commented on app/src/main/java/com/example/findu/presentation/ui/report/component/ReportDateComponent.kt:
```diff
@@ -0,0 +1,105 @@
+            Row(
+                verticalAlignment = Alignment.CenterVertically
+            ) {
+                Text(
+                    text = stringResource(titleRes),
+                    style = FindUTheme.typography.body2SemiBold14,
+                    color = FindUTheme.colors.gray6,
+                )
+                Text(
+                    text = stringResource(R.string.asterisk),
+                    style = FindUTheme.typography.body2SemiBold14.copy(color = FindUTheme.colors.red1)
+                )
```
단순 궁금증인데 color로 텍스트 색상을 지정해주거나 style에서 copy로 색상을 정해주는 기준이 있으신건가요??
👀

ikseong00 replied:
정말 아무 생각 없이 쓰고 있었는데, Text 컴포넌트 내부적으로 색상을 아래처럼 구현하더라구요.
그렇다면 굳이 `copy()` 를 써서 새로운 객체를 생성하는 것보다 color 파라미터로 넘겨주는 편이 나을 것 같습니다.
```kotlin
val localContentColor = LocalContentColor.current
    val localContentAlpha = LocalContentAlpha.current
    val overrideColorOrUnspecified: Color = if (color.isSpecified) {
        color
    } else if (style.color.isSpecified) {
        style.color
    } else {
        localContentColor.copy(localContentAlpha)
    }
```

리뷰 감사합니다!!

ikseong00 replied:
[ee4df71](https://github.com/FindYou-Kuit/FindYou-Android/pull/82/commits/ee4df71d6a4652effa65ee21b90a6e205cfa8724)
마침 다른 것들도 반영했습니다.
Onboarding 쪽도 있었는데 건들지 않았어용

t1nm1ksun reviewed (commented):
1차 리뷰입니다!
회의 하고 또 마저 달게용

ikseong00 commented:
중간부터 커밋 넘버가 #87 들어가있네요 ... 죄송합니다 ..

ikseong00 committed:
[refactor]: #79 앱 설정 다이얼로그 버튼 UI 수정

- RectangleFindUButton 컴포저블 함수 추가
- 앱 설정 다이얼로그 버튼 UI 수정

ikseong00 committed:
[refactor] #79 TopAppBar의 navigationIcon 클릭 핸들러 변경

- TopAppBar 컴포저블에서 `onNavigationIconClick` 매개변수가 `() -> Unit` 타입의 고차 함수를 직접 참조하도록 수정했습니다.
- 기존에는 람다식 내부에서 `onNavigationIconClick()`을 호출했으나, 이제는 직접 참조하여 불필요한 람다 생성을 제거했습니다.

ikseong00 committed:
[chore] #79 날짜 선택 바텀 시트 UI 수정

- 불필요한 import문 삭제
- Modifier 파라미터 위치 수정

ikseong00 committed:
[chore] #79 Text 컴포저블 color 파라미터 분리

- Text 컴포저블의 style 파라미터에 color를 포함하지 않고, color 파라미터로 분리

ksg1227 commented:
@ikseong00 

안녕하세요 심심해서 안드레포 구경왔다가 한 가지 제안드릴게 있어서 댓글 남깁니다.

지금 지도상에서 위치를 지정하고 도로명 주소를 받아오는 과정에서, 도로명 주소가 '서울특별시' 이런 식으로 풀네임이 아니라 '서울' 과 같이 축약된 형태로 받아와지는 것 같아요.

근데 이런 형태 그대로 값을 저장하게 되면, 저희 DB 쪽에 저장해두고 사용하는 시/도 정보랑 형식 불일치하게되어서 글 조회 기능에서 주소 기반 필터링을 할 때 제대로 필터링이 되지 않을 것 같습니다.

그래서 '서울' -> '서울특별시' , '강원' -> '강원특별자치도' 이런 식으로 변환해줄 필요가 있을 것 같은데 어떻게 생각하시나용

ikseong00 commented:
> @ikseong00
> 
> 안녕하세요 심심해서 안드레포 구경왔다가 한 가지 제안드릴게 있어서 댓글 남깁니다.
> 
> 지금 지도상에서 위치를 지정하고 도로명 주소를 받아오는 과정에서, 도로명 주소가 '서울특별시' 이런 식으로 풀네임이 아니라 '서울' 과 같이 축약된 형태로 받아와지는 것 같아요.
> 
> 근데 이런 형태 그대로 값을 저장하게 되면, 저희 DB 쪽에 저장해두고 사용하는 시/도 정보랑 형식 불일치하게되어서 글 조회 기능에서 주소 기반 필터링을 할 때 제대로 필터링이 되지 않을 것 같습니다.
> 
> 그래서 '서울' -> '서울특별시' , '강원' -> '강원특별자치도' 이런 식으로 변환해줄 필요가 있을 것 같은데 어떻게 생각하시나용

@ksg1227 
음 이해했습니다
주소 Api 의 특성상 카카오 Api에서 가져오는 주소는 바뀔 가능 성이 없을 것 같은데(있긴 할 것 같아요),
혹시 그 외에 변동성이 있을까요 ?? 
이 부분에 따라 안드에서 구현할 지 서버에서 구현할 지 정해질 것 같습니다

ksg1227 commented:
> > @ikseong00
> 
> > 
> 
> > 안녕하세요 심심해서 안드레포 구경왔다가 한 가지 제안드릴게 있어서 댓글 남깁니다.
> 
> > 
> 
> > 지금 지도상에서 위치를 지정하고 도로명 주소를 받아오는 과정에서, 도로명 주소가 '서울특별시' 이런 식으로 풀네임이 아니라 '서울' 과 같이 축약된 형태로 받아와지는 것 같아요.
> 
> > 
> 
> > 근데 이런 형태 그대로 값을 저장하게 되면, 저희 DB 쪽에 저장해두고 사용하는 시/도 정보랑 형식 불일치하게되어서 글 조회 기능에서 주소 기반 필터링을 할 때 제대로 필터링이 되지 않을 것 같습니다.
> 
> > 
> 
> > 그래서 '서울' -> '서울특별시' , '강원' -> '강원특별자치도' 이런 식으로 변환해줄 필요가 있을 것 같은데 어떻게 생각하시나용
> 
> 
> 
> @ksg1227 
> 
> 음 이해했습니다
> 
> 주소 Api 의 특성상 카카오 Api에서 가져오는 주소는 바뀔 가능 성이 없을 것 같은데(있긴 할 것 같아요),
> 
> 혹시 그 외에 변동성이 있을까요 ?? 
> 
> 이 부분에 따라 안드에서 구현할 지 서버에서 구현할 지 정해질 것 같습니다

@ikseong00 

서버쪽에서 관리하고있는 정보에도 변동성은 없을 것 같아요. 
서버에서 사용중인 공공데이터 API 에서 시/도 정보를 반환해주는 API 가 존재하는데, V1 진행할 때부터 변동없이 계속 풀네임으로 내려주고있더라구요

아 추가로 구조동물, 분실동물 API 에서 내려주는 응답도 모두 주소를 풀네임으로 내려주고있어유

ikseong00 commented:
> 서버쪽에서 관리하고있는 정보에도 변동성은 없을 것 같아요.
> 서버에서 사용중인 공공데이터 API 에서 시/도 정보를 반환해주는 API 가 존재하는데, V1 진행할 때부터 변동없이 계속 > 풀네임으로 내려주고있더라구요
> 
> 아 추가로 구조동물, 분실동물 API 에서 내려주는 응답도 모두 주소를 풀네임으로 내려주고있어유

@ksg1227 
그러면 안드에서 매핑해서 진행하겠습니다
혹시 정리된 시/도 파일 알려주실 수 있을까용

ksg1227 commented:
> > 서버쪽에서 관리하고있는 정보에도 변동성은 없을 것 같아요.
> 
> > 서버에서 사용중인 공공데이터 API 에서 시/도 정보를 반환해주는 API 가 존재하는데, V1 진행할 때부터 변동없이 계속 > 풀네임으로 내려주고있더라구요
> 
> > 
> 
> > 아 추가로 구조동물, 분실동물 API 에서 내려주는 응답도 모두 주소를 풀네임으로 내려주고있어유
> 
> 
> 
> @ksg1227 
> 
> 그러면 안드에서 매핑해서 진행하겠습니다
> 
> 혹시 정리된 시/도 파일 알려주실 수 있을까용

@ikseong00 
찾아유 백엔드 스페이스 -> AWS 설정 정보 & 더미데이터 ... -> 시도/시군구 데이터 페이지로 가보시면 있습니당

ikseong00 committed:
[feat] #79 주소 정규화 기능 추가

- 주소 문자열 정규화 확장 함수 추가
    - `String.toNormalizeAddress()`: 주소 문자열의 첫 단어를 시/도 전체 이름으로 변환
    - `String.toCityOrProvince()`: 축약된 시/도 이름을 전체 이름으로 변환 (예: "서울" -> "서울특별시")
- 실종/목격 동물 신고 ViewModel에 주소 정규화 로직 적용
    - 신고 완료 버튼 클릭 시, 입력된 주소를 정규화하여 사용

ikseong00 commented:
@ksg1227 
> @ikseong00 찾아유 백엔드 스페이스 -> AWS 설정 정보 & 더미데이터 ... -> 시도/시군구 데이터 페이지로 가보시면 있습니당

1009a9b 반영했습니다!

ksg1227 commented:
> @ksg1227 
> 
> > @ikseong00 찾아유 백엔드 스페이스 -> AWS 설정 정보 & 더미데이터 ... -> 시도/시군구 데이터 페이지로 가보시면 있습니당
> 
> 
> 
> 1009a9b 반영했습니다!
> 
> 

@ikseong00 
아주 굿입니다👍🏻

네이버 지도랑 카카오 지도간에도 주소 표기방법이 다른게 신기하네요.
네이버 지도는 주소룰 전부다 두자리로 표기하는데 (ex) 강원, 세종, 전북, 제주) 카카오는 특정 주소는 풀네임으로 표기하네요.
신기하넹

쨋든 귯

nasohee reviewed (approved):
커밋 단위가 잘 나뉘어져있어서 보기 좋은거같아요 제가 너무 통커밋으로 올린거같다는 생각이 드네요.... 컴포즈 코드도 잘 배워갑니다!
